Source/Dafny/DafnyAst.cs

```diff
@@ -2190,8 +2190,8 @@ public class IntType : BasicType
     public override string TypeName(ModuleDefinition context, bool parseAble) {
       return "int";
     }
-    public override bool Equals(Type that) {
-      return that.IsIntegerType;
+    public override bool Equals(Type that, bool keepConstraints = false) {
+      return keepConstraints ? this.GetType() == that.GetType() : that.IsIntegerType;
```

Comment:
To handle a type like `type MyInt = x: int | x != 8`, I think this needs to be:

    return that.NomalizeExpand(keepConstraints) is IntType;

Reply:
OK